### PR TITLE
Refactors the spider charge

### PR DESCRIPTION
**Pull request by SuperNovaa41**
## About The Pull Request
Refactors the spider charge to track the planting location on itself, rather than constantly accessing the planter's mind (and antag datums)
Also some small code improvements and documentation

## Changelog

:cl:
refactor: The spider charge has slightly less shit-code now.
/:cl:

### CODE REVIEW

**Comment by SuperNovaa41** [comment]
image
he didn't even WAIT for me to do it

File: code/modules/ninja/ninja_explosive.dm [hunk 1]
```diff
@@ -10,38 +10,64 @@
 	name = "spider charge"
 	desc = "A modified C-4 charge supplied to you by the Spider Clan.  Its explosive power has been juiced up, but only works in one specific area."
 	boom_sizes = list(4, 8, 12)
+	///The mob that has planted the charge
 	var/mob/detonator = null
```
Comment (MrMelbert) [inline code/modules/ninja/ninja_explosive.dm:14]: This should probably be a weakref

File: code/modules/ninja/ninja_explosive.dm [hunk 2]
```diff
@@ -10,38 +10,64 @@
...
-	if (!check_loc(user, ninja_antag))
+	var/datum/objective/plant_explosive/objective = locate() in ninja_antag.objectives
+	if (!objective)
+		return
+	detonation_area = objective.detonation_location
+
+/obj/item/grenade/c4/ninja/Destroy()
+	. = ..()
+	detonator = null
+	detonation_area = null
```
Comment (MrMelbert) [inline code/modules/ninja/ninja_explosive.dm:32]:
```suggestion
/obj/item/grenade/c4/ninja/Destroy()
	detonator = null
	detonation_area = null
	return ..()
```

File: code/modules/ninja/ninja_explosive.dm [hunk 3]
```diff
@@ -10,38 +10,64 @@
...
+	///The mob that has planted the charge
 	var/mob/detonator = null
+	///The only area that the charge is allowed to be planted, and detonated in
+	var/detonation_area = null
 
-/obj/item/grenade/c4/ninja/afterattack(atom/movable/AM, mob/user, flag)
-	var/datum/antagonist/ninja/ninja_antag = user.mind.has_antag_datum(/datum/antagonist/ninja)
+/obj/item/grenade/c4/ninja/Initialize(mapload)
+	. = ..()
+	var/mob/living/user = loc
```
Comment (MrMelbert) [inline code/modules/ninja/ninja_explosive.dm:20]: This seems very fragile, I was hoping that you'd be able to pass in the ninja datum as an arg - or in the post equip of the outfit or something idk

**Review by MrMelbert — CHANGES_REQUESTED** [review]
Preliminary review

File: code/modules/ninja/ninja_explosive.dm [hunk 4]
```diff
@@ -10,41 +10,72 @@
...
+		new_c4.say("Invalid location!")
 		target.cut_overlay(plastic_overlay, TRUE)
 		qdel(src)
 		return
 	. = ..()
 	//Since we already did the checks in afterattack, the denonator must be a ninja with the bomb objective.
 	if(!detonator)
 		return
+	var/mob/ninja = detonator.resolve()
+	var/datum/antagonist/ninja/ninja_antag = ninja.mind.has_antag_datum(/datum/antagonist/ninja)
```
Comment (Mothblocks) [inline code/modules/ninja/ninja_explosive.dm:61]: `weakref.resolve()` will return null if the object is deleted.

You have to factor this in. `if (isnull(ninja)) return` will probably work fine here.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Pull request merged**